### ISSUE TITLE
Flaked Flint dupes when repairing Flint Hatchet

### ISSUE THREAD

**Issue by TheLemein** [Darkosto/SevTech-Ages#105]
During early game of Alpha 3 testing

When putting Flint Hatchet, and Flaked Flint on 2x2 inventory crafting table it duplicates the flaked flint upon repair.

https://prnt.sc/gro4io

image to help show, if that doesnt make sense

**Comment by Darkosto** [comment]
Hey @An-Sar any chance this is something on your end?

Damaged flint hatchet + stack of flaked flint to repair=dupe flaked flint.

**Comment by An-Sar** [comment]
Is there a mod that adds anvil repair items to the crafting grid? Bc in my test world you cannot repair with the flaked flint item directly.

**Comment by Darkosto** [comment]
Alright, it might be @BlakeBr0 on this one then. I'll keep digging into it. The weird part is the hatchet seems to be the only tool this currently dupes with. I'm not sure why that is.

**Comment by BlakeBr0** [comment]
It happens with all the tools that say "crafting item" and use a flake as the repair item. I would assume theres something going on with your container items since I can't seem to find the source anywhere.

**Comment by An-Sar** [comment]
those tools all have container items due to having crafting grid recipes that take tool durability on each use, ie: work blade can remove bark from logs in grid, and so on.

`@Override
    public boolean hasContainerItem(ItemStack stack)
    {
        return stack.getItemDamage() < this.getMaxDamage(stack);
    }

    @Override
    public ItemStack getContainerItem(ItemStack stack)
    {
        ItemStack container = stack.copy();
        container.attemptDamageItem(1, ThreadLocalRandom.current(), null);
        return container;
    }`

**Comment by BlakeBr0** [comment]
It doesn't dupe when you repair an iron knife with an iron ingot, so it has to be the flake thingies
Heres the recipe for the grid repair if you think it's my fault https://github.com/BlakeBr0/PickleTweaks2/blob/master/1.12/src/main/java/com/blakebr0/pickletweaks/feature/crafting/GridRepairRecipe.java

**Comment by An-Sar** [comment]
No worries, I don't think anyone said anything about placing fault. Anyway I will add pickle to my environment a bit later and do some testing.

**Comment by An-Sar** [comment]
so I have a crafting event that prevents the vanilla repair recipes, combining tools, from duplicating the tool due to its container item. For some reason this event is causing the remaining stack to duplicate, I'm not currently sure why that's happening.

the key part is reducing the stack size found in the grid, which could be leaving the Pickle's recipe with an unexpected empty itemstack?

`     Item input_item = input_stack.getItem();
            if (input_item instanceof IToolHandler && input_stack.isItemEqualIgnoreDurability(result_stack)) {
                event.craftMatrix.decrStackSize(i, 1);
            }`

**Comment by BlakeBr0** [comment]
It only dupes flakes. When repairing the iron knife with an iron ingot it works just fine, so it has to be something to do with flakes then I'd assume

**Comment by BlakeBr0** [comment]
Maybe. Not sure where I wouldn't be expecting it though. it seems to dupe after decreasing all the stack sizes by 1 like normal

**Comment by BlakeBr0** [comment]
I bet the issue is just container items in general, I'll just make it so container items cant be repaired. Surprised I didn't think about that ;c

**Comment by An-Sar** [comment]
I just changed the Iron Work Blade repair item to an oredict "plateIron" and got it to duplicate IC2 plates.

**Comment by An-Sar** [comment]
Yeah it's a strange bug

**Comment by Darkosto** [comment]
All fixed up! Thank you!

**Comment by justinrusso** [comment]
Pokertsar on discord is having this issue again on Alpha 10.

**Comment by BobChao87** [comment]
Just cloning some notes from Discord for due diligence: Dupe only occurs when Flint Hatchet is broken.

**Comment by 0xb00b** [comment]
Hey Guys, i'll add some more info for those that are curious, i thought it was a fun excercise to explore this bug and try to find the root-cause so heres the result of my little investigation. 

I got the bug to appear first with the PrimalCore flint hatchet.  (using my 4x4 crafting grid) i noticed that when i put in a stack of 2 flaked flint with a broken hatchet, it would not consume any of the flaked flint. A little while later i noticed that when i put in more than 2 flaked flint, i would get duped flint back. When i put in 5 flint, i got 8 flint back. When i put in 4 flint i got 6 back, etc. 

So what seems to be happening is, the act of repairing will actually consume whatever it should consume (1 flaked flint), but whatever is left is fully duped (incidentally, a funny edge-case here is when you just use 1 flint in the crafting grid, then it will produce 'expected' results because (1-1)*2 = 0 and when you use 2, it will give back 2 so it looks like the repair is free. 

This means that if you put in 32 flaked flint and a **broken** hatchet it will consume 1 flint for the repair process, leaving the stack in the grid with 31 flint and when you take out the repaired flint, commiting to the repair process, the 31 flint will be turned into 62 flint. If you repeat this with the flint in the crafting grid, someone on discord pointed out you get a rather funny [result](https://i.imgur.com/IsC44kz.png)

Anyways, it's pretty clear theres a dupe going on here, for some reason the flint stack (repair item material) is being duped during repairing. Since the history chain of this bug suggested the issue is with PickleTweaks, i created a bare multiforge with just mcforge and PickleTweaks. I soon found out that PrimalCore is the mod that adds the flint tools and PickleTweaks really only just serves to put the tools in a 'broken' state, so they can actually be repaired. 

Now, now we know that somehow repairing the hatchet is broken.. the next question becomes are the other flint (or repair-able) tools also broken in the same way? Turns out they are not.. so this leads me to believe the problem is within PrimalCore specifically with the hatchet.. 

Sadly, PrimalCore is not open source, but that doesnt really disable us from looking at the game logic. All we really need is to find out whats different between for example the Primal:flint_pickaxe (which does not exhibit the dupe behaviour) and the Primal:flint_hatchet or whatever theyre referenced as in the code.

After decompiling the latest PrimalCore.jar, we need to find the Item tool classes that contain the repair logic, if we browse through the packages we can easily [spot the different tools..](https://i.imgur.com/5DAoFml.png) but the flint hatchet doesnt have its own class. A text-search through the code gives us an entry in ModItems.class:

 ``` nmd.primal.core.api.PrimalAPI.Items.FLINT_HATCHET = RegistryHelper.addItem("flint_hatchet", new PrimalAxe(PrimalAPI.ToolMaterials.TOOL_FLINT, -2.0F).func_77656_e(PrimalAPI.ToolMaterials.TOOL_FLINT.func_77997_a() / 2), new String[] { "toolHatchetFlint" });```

Where we see that the hatchet is actually a PrimalAxe. Turns out, the flint axe also exhibits the dupe behaviour, but the other tools (pickaxe, hoe etc) behave like expected. 

This all points to some logic in PrimalAxe vs for example PrimalPickAxe that actually produces the dupe, it seems that PickleTweaks is really only needed to put the item in a state where it can be repaired. Fortunately the code isnt that elaborated for both the [PrimalPickAxe](https://pastebin.com/xA1Fvnws)  and [PrimalAxe](https://pastebin.com/XhFn2Yz8) classes. In fact, if we strip out all the code that is the same and other things that cant possibly have to do with anything, were left with this code:

PrimalPickAxe

```public class PrimalPickAxe
  extends ItemPickaxe
  implements IFireProof<ItemTool>
{
  private boolean is_fire_proof;
  
  public PrimalPickAxe(Item.ToolMaterial material)
  {
    super(material);
  }
  
```

PrimalAxe

```

public class PrimalAxe extends ItemAxe implements IToolHandler, IFireProof<ItemAxe>
{
  public PrimalAxe(Item.ToolMaterial material, float speed)
  {
    super(material, material.func_78000_c(), speed);
    
    this.field_77865_bY = (2.0F + material.func_78000_c());
    this.field_185065_c = speed;
  }

  
  public boolean hasContainerItem(ItemStack stack)
  {
    return stack.func_77952_i() < getMaxDamage(stack);
  }
  
  public ItemStack getContainerItem(ItemStack stack)
  {
    ItemStack container = stack.func_77946_l();
    container.func_96631_a(1, PrimalAPI.getRandom(), null);
    return container;
  }
  

}
```
**tl;dr** my best bet is that the bug is somewhere in this trimmed down version of PrimalAxe, someone with actual knowledge of the mcforge api might be able to figure it out. my best bet is on the getContainerItem returning a container but i honestly have no clue maybe im completely wrong and i just wasted 10 minutes of your life :+1:

**Comment by justinrusso** [comment]
Blacklisted flint hatchet. Darkosto/SevTechAgesConfigs@03be8488dc367c84f70dccd2a68f968bb8ee2b96
